contracts/BatchExchangeViewer.sol

```diff
@@ -111,14 +112,17 @@ contract BatchExchangeViewer {
         uint16[] memory tokenFilter,
         address previousPageUser,
         uint16 previousPageUserOffset,
-        uint16 pageSize
-    ) public view returns (bytes memory elements, address nextPageUser, uint16 nextPageUserOffset) {
+        uint16 maxPageSize
+    ) public view returns (bytes memory elements, bool hasNextPage, address nextPageUser, uint16 nextPageUserOffset) {
         nextPageUser = previousPageUser;
         nextPageUserOffset = previousPageUserOffset;
-        bool hasNextPage = true;
-        while (hasNextPage) {
-            bytes memory unfiltered = getEncodedOrdersPaginated(nextPageUser, nextPageUserOffset, pageSize);
-            hasNextPage = unfiltered.length / AUCTION_ELEMENT_WIDTH == pageSize;
+        hasNextPage = true;
+        uint256 gasLeftBeforePage = gasleft();
+        // Continue while more pages exist or we used more than 1/2 of remaining gas in previous page
+        while (hasNextPage && 2 * gasleft() > gasLeftBeforePage) {
+            gasLeftBeforePage = gasleft();
```

Comment:
ubernit: Shouldn't this go after the for loop, in case it didn't return?

Reply:
But then I would compare the gasLeft at the end of the for loop and at the beginning of the while loop, but I need to do it over both loops.

I wasn't easily able to do this without introducing another temp variable (which would push me over the stack too deep limit)

Reply:
Good point!